### ISSUE TITLE
Uusi oikeustaso scriptistä tallentamiseen (kenttä johon voi kirjoittaa vain jsrunnerilla)

### ISSUE THREAD

**Issue by dezhidki** [TIM-JYU/TIM#1450]
In GitLab by @vesal on Jul 19, 2019, 14:58

jsrunnereita voisi joskus antaa ajaa opiskelijan puolellakin.
Muutin oikeuksia niin, että jos komponettilla on `showInView = true`,
niin se saa hakea taulukkoon ja ajaa jsrunneria.  Mutta ei
saa tallentaa jos kenttä on `readonly="view"`.

Kuitenkin voisi olla tilanteita, joissa opiskeija saisi tallentaa kenttää VAIN
skriptin ajamalla. Ja silloinkin pitäisi tarkistaa että skripti on opettajan
kirjoittama.  Esimerkiksi demosumma, kurssin pisteiden summa jne funktiot
voisivat olla tällaisia.  Jos opettaja ei muista painaa Laske, niin opiskelija
ihmettelee miksi kentän arvot ovat väärin.  Silloin voisi laskennan suorittaa
itse (nytkin tuon voi tehdä, mutta silloin kentät on avattava view:ssä kirjoitettaviksi
ja silloin niihin voi kirjoittaa mitä tahansa).  Eli jos tuossa kenttä olisi readonly kuten nytkin,
mutta siinä voisi sanoa että `allowTeacherWrittenJSRunnerSave: true` niin opiskelia voisi
ajaa itse skriptin. Itse asiassa jos tuon voi luotettavasti tarkistaa että on opettajan
kirjoittama skripti, niin miksei jopa voisi olla oletuksena tuo päällä.  Jos jsrunnerin
päästää näkymään View-puolella, on sillä varmaan joku tarkoitus.  Ja jos skriptissä
tallennetaan arvoja, niin silläkin on tarkoitus.

**Comment by sijualle** [comment]
Laittaisin varmaan jonkun yleisen attribuutin (joku runneronly tms), joka estää tuon syötteen antamisen suoraan. Ihan siis riippumatta readonlyistä ja muista. Fieldeissä se harmaa ulkonäkö (disabled) on tosin kiinni tuosta readonlystä, mutta sitä voinee laittaa näkymään myös jos käyttää tätä uutta attribuuttia.


>Itse asiassa jos tuon voi luotettavasti tarkistaa että on opettajan kirjoittama skripti, niin miksei jopa voisi olla oletuksena tuo päällä. Jos jsrunnerin päästää näkymään View-puolella, on sillä varmaan joku tarkoitus. Ja jos skriptissä tallennetaan arvoja, niin silläkin on tarkoitus.

Tämä periaatteessa on jo tietyllä tavalla olemassa. Jsrunnerilla johonkin kenttään kirjoittaessa on kolme toistensa kumoavaa oikeustarkistusta:
- onko opettajaoikeus kohdekentän dokumenttiin 
tai
- onko kohdekenttä samassa dokukumentissa kuin jsrunner ja jsrunneriin laitettu showInView: true
tai
- onko view-oikeus kohdekentän dokumenttiin ja siinä dokumentissa erikseen asetettu että voi kirjoittaa ulkoisilla jsrunnereilla

Nuo kaksi ylimmäistä saa aikaan sen että aina kun perustilanteessa jsrunnerilla pääsee muokkaamaan jotain kenttää niin on joko opettajaoikeus siihen kenttään tai sitten joku edit-oikeuden haltija on mennyt laittamaan siihen dokuun runnerin jossa on showInView: true. Eli oikeastaan tuon jsrunneriin kohdistuva readonly-opettajaoikeustarkistuksen voisi ehkä jopa ottaa kokonaan pois. Silloin voisi suoraan käyttää readonlya määrittämään kentät joihin kirjoitellaan vain runnerilla, eikä tarvitsisi uusia attribuutteja.

**Comment by vesal** [comment]
Tämän lisäksi jää sellaiset tilanteet kuten SUKOLin tilausten käsittelyssä, jossa haluttaisiin käyttää muidenkin vastauksia hyväksi
laskemaan esimerkiksi montako paikkaa on yhteensä varattuna.  Tuossa esimerkissä tarkistetaan että kaikki on oikein,
mutta sitä voisi periaattessa ajaa jokaisen tilauksen jälkeenkin, jolloin mahdollinen yhteentörmäys huomattaisiin
heti:

https://tim.jyu.fi/teacher/sukol/esimerkit/kokeenvaraus#runnerCorrect

Eli miten voitaisiin antaa oikeuksia joihinkin fieldeihin lukea niiden arvoja vaikka ei olisikaan opettaja.

**Comment by dezhidki** [comment]
> Laittaisin varmaan jonkun yleisen attribuutin (joku runneronly tms), joka estää tuon syötteen antamisen suoraan. 

Juu, ihan järkevä näin alkuun.

> Eli miten voitaisiin antaa oikeuksia joihinkin fieldeihin lukea niiden arvoja vaikka ei olisikaan opettaja.

Lienee oman kortin paikka, mutta siis pluginien vastausten lukemiseen ei pitäisi vaatia aina teacher-oikeutta, vaan seeanswers pitäisi riittää. Näin ei taida vielä olla, eli pitää tutkia.

**Comment by vesal** [comment]
> Lienee oman kortin paikka, mutta siis pluginien vastausten lukemiseen ei pitäisi vaatia aina teacher-oikeutta,
> vaan seeanswers pitäisi riittää. Näin ei taida vielä olla, eli pitää tutkia.

seenanswerskin on liikaa, koska jos tuon antaa dokuille, niin silloinhan
voisi mennä katselemaan niitä kaverin vastauksia.  Nythän ajatuksena
on että vain jsrunnerilla joka on jonkun editoikeuden omaavan tekemä,
voisi arvon hakea.  En tiedä miten runnereita voisi "allekirjoittaa",
niin että tuollaisen runnerin tekijän pitäisi esim olla dokun owner
tai vastaava joka voisi ko oikeudet huolimattomuuksissaan antaa
käyttäjille.  Edit-oikeudellahan ei saa oikeuksia antaa.

Silloin sen sijaan että dokuun annetaan mitään lisäoikeuskai, voisi runnerin
attribuuttiin sanoa että tämän ja tämän kentän arvot saa lukea/kirjoittaa
jos sivuun on oikeus X (tässä SUKOL-tapuksessa) se olisi siis että
esim en8-kenttään annettaisiin read-oikeus jos on view-oikeus dokuun.

**Comment by dezhidki** [comment]
> seenanswerskin on liikaa, koska jos tuon antaa dokuille, niin silloinhan voisi mennä katselemaan niitä kaverin vastauksia. Nythän ajatuksena on että vain jsrunnerilla joka on jonkun editoikeuden omaavan tekemä, voisi arvon hakea. En tiedä miten runnereita voisi "allekirjoittaa", niin että tuollaisen runnerin tekijän pitäisi esim olla dokun owner tai vastaava joka voisi ko oikeudet huolimattomuuksissaan antaa käyttäjille. Edit-oikeudellahan ei saa oikeuksia antaa. 

Joo, tosin JSRunnerillakin jo voi muutenkin katsella vastauksia, eli huolettomasti tehdyllä JSRunnerilla voi vuodattaa vastauksia kaikille. Tosiaankin tämän kortin kannalta varmaan voisi katsoa aivan alkuun, saisiko @sijualle:n ehdottaman attribuutin lisäksi myös sellaisen, joka sallii vastausten hakemiseen JSRunneriin samasta dokumentista, jossa itse JSRunner sijaitsee (jotta ei voi vaan tehdä runneria, joka hakee mielivaltaisesta dokumentista vastaukset). Sillä nyt saa ainakin tuon SUKOL-homman hoidettua.

***

Pidemmän päälle vähän epäileväisesti siis suhtaudun siihen, että vastausten käsittelyä käyttäjien tekemillä skripteillä ei suojattaisi oikeudenhallinnalla. Tuo liittyy siinä mielessä tuohon "allekirjoitusjuttuun", että siinäkin pitäisi pystyä tarkemmin hallita, kuka saa ajaa milläkin oikeuksia plugineja. Tuota varten on itse asiassa #463, eli kappalekohtainen oikeushallinta.
